src/lib/asn1/asn1_oid.cpp

```diff
@@ -194,35 +210,73 @@ void OID::decode_from(BER_Decoder& decoder) {
       throw BER_Bad_Tag("Error decoding OID, unknown tag", obj.tagging());
    }
 
-   const size_t length = obj.length();
-   const uint8_t* bits = obj.bits();
-
-   if(length < 2 && !(length == 1 && bits[0] == 0)) {
+   if(obj.length() == 0) {
       throw BER_Decoding_Error("OID encoding is too short");
    }
 
-   m_id.clear();
-   m_id.push_back(bits[0] / 40);
-   m_id.push_back(bits[0] % 40);
+   auto consume = [](std::span<const uint8_t> data) -> std::pair<std::span<const uint8_t>, uint32_t> {
```

Comment:
IMHO this is a good use case for a `BufferSlicer&` and letting `consume` call `BufferSlicer::take_byte()` as often as needed.  That would remove most of the buffer bookkeeping from the implementation in `consume` and also simplify the return type to just a `uint32_t`.